python-lib/dku_email_client.py

```diff
@@ -0,0 +1,112 @@
+
+
+class AbstractMessageClient(ABC):
+    def __init__(self, plain_text):
+        self.plain_text = plain_text
+
+
+    def send_email(self, sender, recipient, email_subject, email_body, attachment_files):
+        """
+        :param sender: sender email, str - depending on impl, could be ignored if a sender configured for the channel
+        :param recipient: recipient email, str
+        :param email_subject: str
+        :param email_body: body of either plain text or html, str
+        :param attachment_files:attachments as list of  AttachmentFile
+        """
+        # Any generic email body decoration goes here
+        email_body_to_send = email_body
+        if self.plain_text:
+            email_body_to_send = email_body + '\n\n'
+
+        self.send_email_impl(sender, recipient, email_subject, email_body_to_send, attachment_files)
+
+    @abstractmethod
+    def send_email_impl(self,  sender, recipient, email_body, email_subject, attachment_files):
+        pass
+
+
+class ChannelClient(AbstractMessageClient):
+    """ Impl using DSS channels that requires DSS 12.6 or later """
+    def __init__(self, plain_text, channel_id):
+        super().__init__(plain_text)
+
+        self.dss_client = dataiku.api_client()
+        self.project_id = dataiku.default_project_key()
+        self.channel = self.dss_client.get_messaging_channel(channel_id)
+
+        logging.info(f"Configured channel messaging client with channel {channel_id} - type: {self.channel.type}, "
+                     f"sender: {self.channel.sender}, plain_text? {self.plain_text}")
+
+    def send_email_impl(self, sender, recipient, email_subject, email_body, attachment_files):
+        """
+        :param sender: sender email, str - is ignored if a sender configured for the channel
+        :param recipient: recipient email, str
+        :param email_subject: str
+        :param email_body: body of either plain text or html, str
+        :param attachment_files:attachments as list of  AttachmentFile
+        """
+        files = [(a.file_name, a.data, f"{a.mime_type}/{a.mime_subtype}") for a in attachment_files]
+
+        sender_to_use = None if self.channel.sender else sender
+        self.channel.send(self.project_id, [recipient], email_subject, email_body, attachments=files, plain_text=self.plain_text, sender=sender_to_use)
+
+    def quit(self):
+        pass
+
+
+class SmtpEmailClient(AbstractMessageClient):
+    """ Client for sending email - direct SMTP implementation """
+
+    def __init__(self, plain_text, smtp_config):
+        super().__init__(plain_text)
+        self.smtp = smtplib.SMTP(smtp_config.smtp_host, port=smtp_config.smtp_port)
+        # Use TLS if set
+        if smtp_config.smtp_use_tls:
+            self.smtp.starttls()
+        # Use credentials if set
+        if smtp_config.smtp_use_auth:
+            self.smtp.login(str(smtp_config.smtp_user), str(smtp_config.smtp_pass))
```

Comment:
all the init should be performed in a dedicated method, "performLogin" for instance

Reply:
Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/d66b102869e15faa70d737a5e1df68c31b08a8cb)